packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -10,39 +10,112 @@ import { ThrottlingError } from "@fluidframework/driver-utils";
+ } from "./odspCache";
 import { RateLimiter } from "./rateLimiter";
 
-export type FetchType = "blob" | "createBlob" | "createFile" | "joinSession" | "ops" | "other" | "snapshotTree" |
+export type FetchType = "blob" | "createBlob" | "createFile" | "joinSession" | "ops" | "test" | "snapshotTree" |
     "treesLatest" | "uploadSummary" | "push" | "versions";
 
+export type FetchTypeInternal = FetchType | "cache";
+
+// exported only of test purposes
+export interface IVersionedValueWithEpoch {
+    value: any;
+    fluidEpoch: string,
+    version: 2,
+}
+
+// exported only of test purposes
+export const persistedCacheValueVersion = 2;
+
 /**
  * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
  * server can match it with its epoch value in order to match the version.
  * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
  * then it also clears all the cached entries for the given container.
  */
-export class EpochTracker {
+export class EpochTracker implements IPersistedFileCache {
     private _fluidEpoch: string | undefined;
-    private _fileEntry: IFileEntry | undefined;
+
     public readonly rateLimiter: RateLimiter;
 
     constructor(
-        private readonly persistedCache: LocalPersistentCacheAdapter,
+        protected readonly cache: IPersistedCache,
+        protected readonly fileEntry: IFileEntry,
         protected readonly logger: ITelemetryLogger,
     ) {
         // Limits the max number of concurrent requests to 24.
         this.rateLimiter = new RateLimiter(24);
     }
 
-    public set fileEntry(fileEntry: IFileEntry | undefined) {
-        assert(this._fileEntry === undefined, 0x09b /* "File Entry should be set only once" */);
-        assert(fileEntry !== undefined, 0x09c /* "Passed file entry should not be undefined" */);
-        this._fileEntry = fileEntry;
+    // public for UT purposes only!
+    public setEpoch(epoch: string, fromCache: boolean, fetchType: FetchTypeInternal) {
+        assert(this._fluidEpoch === undefined, "epoch exists");
+        this._fluidEpoch = epoch;
+
+        this.logger.sendTelemetryEvent(
+            {
+                eventName: "EpochLearnedFirstTime",
+                epoch,
+                fetchType,
+                fromCache,
+            },
+        );
+    }
+
+    public async get(
+        entry: IEntry,
+    ): Promise<any> {
+        try {
+            const value: IVersionedValueWithEpoch = await this.cache.get(this.fileEntryFromEntry(entry));
+            if (value === undefined || value.version !== persistedCacheValueVersion) {
+                return undefined;
+            }
+            assert(value.fluidEpoch !== undefined, "all entries have to have epoch");
+            if (this._fluidEpoch === undefined) {
+                this.setEpoch(value.fluidEpoch, true, "cache");
+            } else if (this._fluidEpoch !== value.fluidEpoch) {
+                return undefined;
```

Comment:
On mismatch, should not we throw and clear cache entries?

Reply:
Definitely not throw. Consider the case that all other items in cache expired and it just has one image from old session with old epoch. If we loaded from storage snapshot with new epoch, but got a hit for image, we should just disregard it and move forward.
Deleting it would be nice, but currently caching interface does not have delete per key. I think that's fine, as intention for the cache to be relatively short lived.